File: src/tools/opencode.py
"""Sandbox tool — runs opencode or claude-code in an ephemeral Docker container."""

import asyncio
import os
import shutil
from collections.abc import Awaitable, Callable
from pathlib import Path
from uuid import uuid4
from typing import Any

import docker
import docker.errors
import sentry_sdk

SANDBOX_IMAGE = os.getenv("SANDBOX_IMAGE", "house-chatbot-sandbox:latest")
DOCKER_NETWORK = os.getenv("DOCKER_NETWORK", "house-chatbot_default")
TIMEOUT = int(os.getenv("SANDBOX_TIMEOUT", "300"))
SANDBOX_CPU_QUOTA = int(
    os.getenv("SANDBOX_CPU_QUOTA", "200000")
)  # 2 CPUs (100000 = 1 CPU per 100ms period)
SANDBOX_MEM_LIMIT = os.getenv("SANDBOX_MEM_LIMIT", "1g")
ARTIFACTS_DIR = os.getenv("ARTIFACTS_DIR", "data/artifacts")
MAX_ARTIFACT_SIZE_MB = int(os.getenv("MAX_ARTIFACT_SIZE_MB", "24"))

# HOST_DATA_DIR is the host-side absolute path to ./data, needed because sandbox containers
# are Docker siblings — volume paths must be resolvable by the Docker daemon on the host,
# not inside this container's filesystem.
HOST_DATA_DIR = os.getenv("HOST_DATA_DIR", "")
CONTAINER_DATA_DIR = os.getenv("DATA_DIR", "data")

TOOL_DEFINITION: dict[str, Any] = {
    "name": "run_opencode",
    "description": (
        "Run a software development task using OpenCode powered by a local llama.cpp model. "
        "Good for general coding tasks, quick scripts, and iterative work. "
        "Optionally clone a git repo or seed the workspace with files."
    ),
    "input_schema": {
        "type": "object",
        "properties": {
            "task": {
                "type": "string",
                "description": "The software development task to perform.",
            },
            "model": {
                "type": "string",
                "description": (
                    "Model to use, e.g. server-slop/qwen3.6-35b or server-slop/ornith-1.0-35b. "
                    "Defaults to server-slop/gemma-4-12b-qat-q4kxl."
                ),
            },
            "repo_url": {
                "type": "string",
                "description": "Optional Git repository URL to clone into the workspace.",
            },
            "files": {
                "type": "object",
                "description": "Optional map of relative file paths to content to seed before running.",
                "additionalProperties": {"type": "string"},
            },
        },
        "required": ["task"],
    },
}

ProgressCallback = Callable[[str], Awaitable[None]]


async def run_opencode(
    task: str,
    model: str | None = None,
    repo_url: str | None = None,
    files: dict[str, str] | None = None,
    on_progress: ProgressCallback | None = None,
) -> dict[str, Any] | str:
    with sentry_sdk.start_span(op="sandbox.run", name="run_opencode") as span:
        span.set_data("task", task[:500])
        span.set_data("model", model or "default")
        span.set_data("repo_url", repo_url or "")
        span.set_data("seed_files", list(files.keys()) if files else [])
        return await _call_sandbox(
            "opencode", task, repo_url, files, model=model, on_progress=on_progress
        )


async def _call_sandbox(
    agent: str,
    task: str,
    repo_url: str | None,
    files: dict[str, str] | None,
    model: str | None = None,
    on_progress: ProgressCallback | None = None,
) -> dict[str, Any] | str:
    loop = asyncio.get_running_loop()
    return await loop.run_in_executor(
        None,
        _run_container_sync,
        agent,
        task,
        repo_url,
        files,
        model,
        on_progress,
        loop,
    )


def _make_workspace() -> tuple[str, str]:
    """Create a workspace dir and return (host_path, container_path).

    The host_path is passed to Docker as the volume source (visible to the daemon).
    The container_path is where this process reads/writes the same directory.
    They point to the same filesystem location via the ./data bind mount.
    """
    uid = uuid4().hex[:8]
    container_path = os.path.join(CONTAINER_DATA_DIR, "workspaces", uid)
    os.makedirs(container_path, exist_ok=True)
    os.chmod(container_path, 0o777)
    host_path = (
        os.path.join(HOST_DATA_DIR, "workspaces", uid)
        if HOST_DATA_DIR
        else container_path
    )
    return host_path, container_path


_EXCLUDED_FILENAMES = {"opencode.json", ".opencode.json"}


def _collect_workspace_files(container_workspace: str) -> list[str]:
    """Copy individual workspace files into ARTIFACTS_DIR; return their paths."""
    collected: list[str] = []
    uid = uuid4().hex[:8]
    os.makedirs(ARTIFACTS_DIR, exist_ok=True)

    for root, _dirs, files in os.walk(container_workspace):
        _dirs[:] = [d for d in _dirs if not d.startswith(".")]
        for filename in files:
            if filename in _EXCLUDED_FILENAMES or filename.startswith("."):
                continue
            src = os.path.join(root, filename)
            size_mb = os.path.getsize(src) / (1024 * 1024)
            if size_mb > MAX_ARTIFACT_SIZE_MB:
                continue
            rel = os.path.relpath(src, container_workspace)
            flat = rel.replace(os.sep, "_")
            dst = os.path.join(ARTIFACTS_DIR, f"{uid}_{flat}")
            shutil.copy2(src, dst)
            collected.append(dst)

    return collected


def _run_container_sync(
    agent: str,
    task: str,
    repo_url: str | None,
    files: dict[str, str] | None,
    model: str | None,
    on_progress: ProgressCallback | None,
    loop: asyncio.AbstractEventLoop,
) -> dict[str, Any] | str:
    try:
        client = docker.from_env()
    except docker.errors.DockerException as exc:
        return f"Error: cannot connect to Docker socket: {exc}"

    environment: dict[str, str] = {
        "AGENT": agent,
        "TASK": task,
        "REPO_URL": repo_url or "",
        "MODEL": model or "",
        "NO_COLOR": "1",
        "TERM": "dumb",
    }
    for var in ("LLAMA_CPP_URL", "LLAMA_CPP_MODEL"):
        val = os.getenv(var, "")
        if val:
            environment[var] = val
    cc_token = os.getenv("CC_OAUTH_TOKEN") or os.getenv("CLAUDE_CODE_OAUTH_TOKEN", "")
    if cc_token:
        environment["CLAUDE_CODE_OAUTH_TOKEN"] = cc_token

    host_workspace, container_workspace = _make_workspace()
    try:
        if files:
            for rel, content in files.items():
                p = Path(container_workspace) / rel
                p.parent.mkdir(parents=True, exist_ok=True)
                p.write_text(content)

        try:
            container = client.containers.run(
                SANDBOX_IMAGE,
                name=f"sandbox-{uuid4().hex[:8]}",
                network=DOCKER_NETWORK,
                environment=environment,
                volumes={host_workspace: {"bind": "/workspace", "mode": "rw"}},
                detach=True,
                remove=False,  # we remove manually after streaming logs
                cpu_quota=SANDBOX_CPU_QUOTA,
                cpu_period=100000,
                mem_limit=SANDBOX_MEM_LIMIT,
            )
        except docker.errors.ImageNotFound:
            return f"Error: sandbox image '{SANDBOX_IMAGE}' not found — run: docker compose build sandbox"
        except docker.errors.APIError:
            raise

        lines: list[str] = []
        artifact_paths: list[str] = []
        try:
            for raw in container.logs(stream=True, follow=True):
                line = raw.decode(errors="replace")
                lines.append(line)
                if on_progress:
                    asyncio.run_coroutine_threadsafe(on_progress(line), loop).result(
                        timeout=5
                    )

            result = container.wait(timeout=TIMEOUT)
            exit_code = result.get("StatusCode", -1)
            if exit_code != 0:
                output = "".join(lines).strip()
                return f"Error: sandbox exited with code {exit_code}.\n{output}"

            artifact_paths = _collect_workspace_files(container_workspace)
        except Exception as exc:
            try:
                container.kill()
            except Exception:
                pass
            return f"Error: sandbox failed: {exc}"
        finally:
            try:
                container.remove(force=True)
            except Exception:
                pass
    finally:
        shutil.rmtree(container_workspace, ignore_errors=True)

    output = "".join(lines).strip() or "(no output)"
    result = {"content": output}
    if artifact_paths:
        result["_artifact_paths"] = artifact_paths
    return result

File: src/tools/opencode.rs
//! Sandbox tool — runs `opencode` in an ephemeral Docker container.
//!
//! The container is a Docker *sibling* (it talks to the host daemon), so the workspace
//! is bind-mounted from a host-visible path under `HOST_DATA_DIR`. Log lines stream back
//! to the caller via a [`TextSink`]; individual output files are copied into the artifacts
//! directory and returned for upload.

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::time::Duration;

use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use uuid::Uuid;

use crate::config;
use crate::llm::TextSink;

/// Filenames never treated as user artifacts.
pub const EXCLUDED_FILENAMES: &[&str] = &["opencode.json", ".opencode.json"];

/// OpenAI-style tool definition (internal `input_schema` form).
pub fn definition() -> serde_json::Value {
    serde_json::json!({
        "name": "run_opencode",
        "description": "Run a software development task using OpenCode powered by a local \
            llama.cpp model. Good for general coding tasks, quick scripts, and iterative work. \
            Optionally clone a git repo or seed the workspace with files.",
        "input_schema": {
            "type": "object",
            "properties": {
                "task": {"type": "string", "description": "The software development task to perform."},
                "model": {"type": "string", "description": "Model to use, e.g. server-slop/qwen3.6-35b. Defaults to server-slop/gemma-4-12b-qat-q4kxl."},
                "repo_url": {"type": "string", "description": "Optional Git repository URL to clone into the workspace."},
                "files": {"type": "object", "description": "Optional map of relative file paths to content to seed before running.", "additionalProperties": {"type": "string"}}
            },
            "required": ["task"]
        }
    })
}

/// Result of a sandbox run: streamed output plus any collected artifact files.
#[derive(Debug, Default, Clone)]
pub struct OpencodeOutput {
    pub content: String,
    pub artifact_paths: Vec<PathBuf>,
}

/// Copy eligible workspace files into `artifacts_dir`, returning their new paths.
///
/// Skips dotfiles, dot-directories, [`EXCLUDED_FILENAMES`], and files larger than
/// `max_artifact_mb` megabytes.
pub fn collect_workspace_files(
    workspace: &Path,
    artifacts_dir: &Path,
    max_artifact_mb: f64,
) -> Vec<PathBuf> {
    let mut collected = Vec::new();
    if std::fs::create_dir_all(artifacts_dir).is_err() {
        return collected;
    }
    let uid = short_uid();
    walk(
        workspace,
        workspace,
        artifacts_dir,
        max_artifact_mb,
        &uid,
        &mut collected,
    );
    collected
}

fn walk(
    root: &Path,
    dir: &Path,
    artifacts_dir: &Path,
    max_mb: f64,
    uid: &str,
    out: &mut Vec<PathBuf>,
) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };
    for entry in entries.flatten() {
        let path = entry.path();
        let name = entry.file_name().to_string_lossy().to_string();
        if path.is_dir() {
            if name.starts_with('.') {
                continue; // don't descend into dot-directories
            }
            walk(root, &path, artifacts_dir, max_mb, uid, out);
            continue;
        }
        if name.starts_with('.') || EXCLUDED_FILENAMES.contains(&name.as_str()) {
            continue;
        }
        let size_mb = std::fs::metadata(&path)
            .map(|m| m.len() as f64 / (1024.0 * 1024.0))
            .unwrap_or(f64::MAX);
        if size_mb > max_mb {
            continue;
        }
        let rel = path.strip_prefix(root).unwrap_or(&path);
        let flat = rel
            .to_string_lossy()
            .replace(std::path::MAIN_SEPARATOR, "_");
        let dst = artifacts_dir.join(format!("{uid}_{flat}"));
        if std::fs::copy(&path, &dst).is_ok() {
            out.push(dst);
        }
    }
}

fn short_uid() -> String {
    Uuid::new_v4().simple().to_string()[..8].to_string()
}

// ── runtime configuration ────────────────────────────────────────────────────

struct SandboxConfig {
    image: String,
    network: String,
    timeout: Duration,
    cpus: String,
    mem: String,
    artifacts_dir: PathBuf,
    max_artifact_mb: f64,
    host_data_dir: String,
    container_data_dir: String,
}

impl SandboxConfig {
    fn from_env() -> Self {
        Self {
            image: config::env_or("SANDBOX_IMAGE", "house-chatbot-sandbox:latest"),
            network: config::env_or("DOCKER_NETWORK", "house-chatbot_default"),
            timeout: Duration::from_secs(config::env_parse("SANDBOX_TIMEOUT", 300)),
            cpus: config::env_or("SANDBOX_CPUS", "2"),
            mem: config::env_or("SANDBOX_MEM_LIMIT", "1g"),
            artifacts_dir: PathBuf::from(config::env_or("ARTIFACTS_DIR", "data/artifacts")),
            max_artifact_mb: config::env_parse("MAX_ARTIFACT_SIZE_MB", 24.0),
            host_data_dir: config::env_or("HOST_DATA_DIR", ""),
            container_data_dir: config::env_or("DATA_DIR", "data"),
        }
    }
}

/// Run an OpenCode task in a fresh sandbox container.
pub async fn run_opencode(
    task: &str,
    model: Option<&str>,
    repo_url: Option<&str>,
    files: Option<&HashMap<String, String>>,
    sink: Option<&dyn TextSink>,
) -> OpencodeOutput {
    let cfg = SandboxConfig::from_env();

    let uid = short_uid();
    let host_workspace = if cfg.host_data_dir.is_empty() {
        PathBuf::from(&cfg.container_data_dir)
            .join("workspaces")
            .join(&uid)
    } else {
        PathBuf::from(&cfg.host_data_dir)
            .join("workspaces")
            .join(&uid)
    };
    if let Err(e) = std::fs::create_dir_all(&host_workspace) {
        return err(format!("cannot create workspace: {e}"));
    }

    // Seed files.
    if let Some(files) = files {
        for (rel, content) in files {
            let p = host_workspace.join(rel);
            if let Some(parent) = p.parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            let _ = std::fs::write(&p, content);
        }
    }

    let container_name = format!("sandbox-{}", short_uid());
    let mut cmd = Command::new("docker");
    cmd.arg("run")
        .arg("--rm")
        .arg("--name")
        .arg(&container_name)
        .arg("--network")
        .arg(&cfg.network)
        .arg("-e")
        .arg("AGENT=opencode")
        .arg("-e")
        .arg(format!("TASK={task}"))
        .arg("-e")
        .arg(format!("REPO_URL={}", repo_url.unwrap_or("")))
        .arg("-e")
        .arg(format!("MODEL={}", model.unwrap_or("")))
        .arg("-e")
        .arg("NO_COLOR=1")
        .arg("-e")
        .arg("TERM=dumb");
    for var in ["LLAMA_CPP_URL", "LLAMA_CPP_MODEL"] {
        if let Ok(val) = std::env::var(var) {
            if !val.is_empty() {
                cmd.arg("-e").arg(format!("{var}={val}"));
            }
        }
    }
    cmd.arg("--cpus")
        .arg(&cfg.cpus)
        .arg("--memory")
        .arg(&cfg.mem)
        .arg("-v")
        .arg(format!("{}:/workspace:rw", host_workspace.display()))
        .arg(&cfg.image)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let result = run_streaming(cmd, cfg.timeout, &container_name, sink).await;

    let output = match result {
        Ok((code, lines)) => {
            if code != 0 {
                let _ = std::fs::remove_dir_all(&host_workspace);
                return err(format!(
                    "sandbox exited with code {code}.\n{}",
                    lines.trim()
                ));
            }
            lines
        }
        Err(e) => {
            let _ = tokio::process::Command::new("docker")
                .arg("kill")
                .arg(&container_name)
                .output()
                .await;
            let _ = std::fs::remove_dir_all(&host_workspace);
            return err(format!("sandbox failed: {e}"));
        }
    };

    let artifacts =
        collect_workspace_files(&host_workspace, &cfg.artifacts_dir, cfg.max_artifact_mb);
    let _ = std::fs::remove_dir_all(&host_workspace);

    let content = {
        let t = output.trim();
        if t.is_empty() {
            "(no output)".to_string()
        } else {
            t.to_string()
        }
    };
    OpencodeOutput {
        content,
        artifact_paths: artifacts,
    }
}

fn err(msg: String) -> OpencodeOutput {
    OpencodeOutput {
        content: format!("Error: {msg}"),
        artifact_paths: vec![],
    }
}

/// Spawn `cmd`, stream merged stdout/stderr lines to `sink`, and return (exit_code, full_output).
async fn run_streaming(
    mut cmd: Command,
    timeout: Duration,
    container_name: &str,
    sink: Option<&dyn TextSink>,
) -> anyhow::Result<(i32, String)> {
    let mut child = cmd.spawn()?;
    let stdout = child.stdout.take();
    let stderr = child.stderr.take();

    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<String>();
    if let Some(out) = stdout {
        let tx = tx.clone();
        tokio::spawn(async move {
            let mut lines = BufReader::new(out).lines();
            while let Ok(Some(line)) = lines.next_line().await {
                let _ = tx.send(line);
            }
        });
    }
    if let Some(err) = stderr {
        let tx = tx.clone();
        tokio::spawn(async move {
            let mut lines = BufReader::new(err).lines();
            while let Ok(Some(line)) = lines.next_line().await {
                let _ = tx.send(line);
            }
        });
    }
    drop(tx);

    let mut collected = String::new();
    let collect = async {
        while let Some(line) = rx.recv().await {
            collected.push_str(&line);
            collected.push('\n');
            if let Some(s) = sink {
                s.push(&line).await;
            }
        }
        let status = child.wait().await?;
        Ok::<i32, anyhow::Error>(status.code().unwrap_or(-1))
    };

    match tokio::time::timeout(timeout, collect).await {
        Ok(code) => Ok((code?, collected)),
        Err(_) => {
            let _ = Command::new("docker")
                .arg("kill")
                .arg(container_name)
                .output()
                .await;
            anyhow::bail!("timed out after {}s", timeout.as_secs())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn setup() -> (TempDir, PathBuf, PathBuf) {
        let tmp = TempDir::new().unwrap();
        let ws = tmp.path().join("workspace");
        std::fs::create_dir_all(&ws).unwrap();
        let artifacts = tmp.path().join("artifacts");
        (tmp, ws, artifacts)
    }

    #[test]
    fn empty_workspace_returns_empty() {
        let (_t, ws, art) = setup();
        assert!(collect_workspace_files(&ws, &art, 24.0).is_empty());
    }

    #[test]
    fn regular_file_is_collected() {
        let (_t, ws, art) = setup();
        std::fs::write(ws.join("hello.py"), "print('hi')").unwrap();
        let r = collect_workspace_files(&ws, &art, 24.0);
        assert_eq!(r.len(), 1);
        assert!(r[0].to_string_lossy().contains("hello"));
    }

    #[test]
    fn excluded_opencode_json_is_skipped() {
        let (_t, ws, art) = setup();
        std::fs::write(ws.join("opencode.json"), "{}").unwrap();
        assert!(collect_workspace_files(&ws, &art, 24.0).is_empty());
    }

    #[test]
    fn excluded_dot_opencode_json_is_skipped() {
        let (_t, ws, art) = setup();
        std::fs::write(ws.join(".opencode.json"), "{}").unwrap();
        assert!(collect_workspace_files(&ws, &art, 24.0).is_empty());
    }

    #[test]
    fn dotfile_is_skipped() {
        let (_t, ws, art) = setup();
        std::fs::write(ws.join(".gitignore"), "*.pyc").unwrap();
        assert!(collect_workspace_files(&ws, &art, 24.0).is_empty());
    }

    #[test]
    fn dotdir_contents_are_skipped() {
        let (_t, ws, art) = setup();
        let hidden = ws.join(".git");
        std::fs::create_dir_all(&hidden).unwrap();
        std::fs::write(hidden.join("config"), "data").unwrap();
        assert!(collect_workspace_files(&ws, &art, 24.0).is_empty());
    }

    #[test]
    fn oversized_file_is_skipped() {
        let (_t, ws, art) = setup();
        std::fs::write(ws.join("big.txt"), "data").unwrap();
        assert!(collect_workspace_files(&ws, &art, 0.0).is_empty());
    }

    #[test]
    fn multiple_files_all_collected() {
        let (_t, ws, art) = setup();
        std::fs::write(ws.join("a.py"), "a").unwrap();
        std::fs::write(ws.join("b.sh"), "b").unwrap();
        assert_eq!(collect_workspace_files(&ws, &art, 24.0).len(), 2);
    }

    #[test]
    fn nested_file_is_collected() {
        let (_t, ws, art) = setup();
        let sub = ws.join("src");
        std::fs::create_dir_all(&sub).unwrap();
        std::fs::write(sub.join("main.py"), "code").unwrap();
        assert_eq!(collect_workspace_files(&ws, &art, 24.0).len(), 1);
    }

    #[test]
    fn collected_files_exist_in_artifacts_dir() {
        let (_t, ws, art) = setup();
        std::fs::write(ws.join("out.txt"), "output").unwrap();
        let r = collect_workspace_files(&ws, &art, 24.0);
        assert_eq!(r.len(), 1);
        assert!(r[0].is_file());
    }

    #[test]
    fn excluded_filenames_constant() {
        assert!(EXCLUDED_FILENAMES.contains(&"opencode.json"));
        assert!(EXCLUDED_FILENAMES.contains(&".opencode.json"));
    }

    #[test]
    fn definition_requires_task() {
        assert_eq!(
            definition()["input_schema"]["required"],
            serde_json::json!(["task"])
        );
    }
}
